compiler/noirc_frontend/src/parser/parser.rs

```diff
@@ -91,1764 +44,504 @@ use traits::where_clause;
-            atom_or_right_unary(expression(), expression_nc, fresh_statement(), true, parse_type()),
-            vec!["x as pub u8"],
-        );
+    fn eat_str(&mut self) -> Option<String> {
+        if matches!(self.token.token(), Token::Str(..)) {
+            let token = std::mem::take(&mut self.token);
+            self.next_token();
+            match token.into_token() {
+                Token::Str(string) => Some(string),
+                _ => unreachable!(),
+            }
+        } else {
+            None
+        }
     }
 
-    #[test]
-    fn parse_array_index() {
-        let valid = vec![
-            "x[9]",
-            "y[x+a]",
-            " foo [foo+5]",
-            "baz[bar]",
-            "foo.bar[3] as Field .baz as u32 [7]",
-        ];
-        parse_all(
-            atom_or_right_unary(
-                expression(),
-                expression_no_constructors(expression()),
-                fresh_statement(),
-                true,
-                parse_type(),
-            ),
-            valid,
-        );
+    fn eat_raw_str(&mut self) -> Option<(String, u8)> {
+        if matches!(self.token.token(), Token::RawStr(..)) {
+            let token = std::mem::take(&mut self.token);
+            self.next_token();
+            match token.into_token() {
+                Token::RawStr(string, n) => Some((string, n)),
+                _ => unreachable!(),
+            }
+        } else {
+            None
+        }
     }
 
-    fn expr_to_array(expr: ExpressionKind) -> ArrayLiteral {
-        let lit = match expr {
-            ExpressionKind::Literal(literal) => literal,
-            _ => unreachable!("expected a literal"),
-        };
-
-        match lit {
-            Literal::Array(arr) => arr,
-            _ => unreachable!("expected an array"),
+    fn eat_fmt_str(&mut self) -> Option<String> {
+        if matches!(self.token.token(), Token::FmtStr(..)) {
+            let token = std::mem::take(&mut self.token);
+            self.next_token();
+            match token.into_token() {
+                Token::FmtStr(string) => Some(string),
+                _ => unreachable!(),
+            }
+        } else {
+            None
         }
     }
 
-    #[test]
-    fn parse_array() {
-        let valid = vec![
-            "[0, 1, 2,3, 4]",
-            "[0,1,2,3,4,]", // Trailing commas are valid syntax
-            "[0;5]",
-        ];
-
-        for expr in parse_all(array_expr(expression()), valid) {
-            match expr_to_array(expr) {
-                ArrayLiteral::Standard(elements) => assert_eq!(elements.len(), 5),
-                ArrayLiteral::Repeated { length, .. } => {
-                    assert_eq!(length.kind, ExpressionKind::integer(5i128.into()));
-                }
+    fn eat_quote(&mut self) -> Option<Tokens> {
+        if matches!(self.token.token(), Token::Quote(..)) {
+            let token = std::mem::take(&mut self.token);
+            self.next_token();
```

Comment:
If `self.next_token()` returned the previous token we wouldn't need to `take`. Could be a bit confusing though that it could look like it was returning the next token. Maybe it'd be solved with a rename to `self.advance()` or similar but I'm just nitpicking now.